--- pluginloader.cpp	2017-07-06 10:17:19.373575300 -0700
+++ pluginloader.cpp	2017-07-06 11:06:16.320507700 -0700
@@ -25,6 +25,7 @@
 
 #include <stdlib.h>
 
+#include <QDebug>
 #include <QtGlobal>
 #include <QtPlugin>
 #include <QCoreApplication>
@@ -37,6 +38,181 @@
 #include "../qpy/QtDesigner/qpydesignercustomwidgetplugin.h"
 
 
+const char * moduleName = "exceptiondialog";
+const char * moduleCode = "\n\
+import functools\n\
+import sys\n\
+import traceback\n\
+\n\
+from PyQt5 import QtCore, QtWidgets\n\
+\n\
+QtCore.qDebug('Importing module for exception dialogs')\n\
+\n\
+def exception_message_box(excType=None, excValue=None, tracebackobj=None):\n\
+    message = traceback.format_exception(excType, excValue, tracebackobj)\n\
+    message = ''.join(message)\n\
+\n\
+    dialog(\n\
+        parent=None,\n\
+        title='Exception',\n\
+        message=message,\n\
+        icon=QtWidgets.QMessageBox.Critical,\n\
+    )\n\
+\n\
+class FittedTextBrowser(QtWidgets.QTextBrowser):\n\
+    def sizeHint(self):\n\
+        default = super().sizeHint()\n\
+\n\
+        if not default.isValid():\n\
+            return default\n\
+\n\
+        document_size = self.document().size()\n\
+\n\
+        desktops = QtWidgets.QApplication.desktop()\n\
+        screen_number = desktops.screenNumber(self.parent())\n\
+        geometry = desktops.screenGeometry(screen_number)\n\
+\n\
+        if document_size.width() == 0:\n\
+            document_size.setWidth(geometry.width() * 0.25)\n\
+        if document_size.height() == 0:\n\
+            document_size.setHeight(geometry.height() * 0.4)\n\
+\n\
+        scrollbar_width = QtWidgets.QApplication.style().pixelMetric(\n\
+            QtWidgets.QStyle.PM_ScrollBarExtent\n\
+        )\n\
+\n\
+        width = sum((\n\
+            document_size.width(),\n\
+            self.contentsMargins().left(),\n\
+            self.contentsMargins().right(),\n\
+            scrollbar_width,\n\
+        ))\n\
+\n\
+        height = sum((\n\
+            document_size.height(),\n\
+            self.contentsMargins().top(),\n\
+            self.contentsMargins().bottom(),\n\
+            scrollbar_width,\n\
+        ))\n\
+\n\
+        return QtCore.QSize(width, height)\n\
+\n\
+\n\
+class DialogUi:\n\
+    def __init__(self, parent):\n\
+        self.layout = QtWidgets.QGridLayout(parent)\n\
+        self.icon = QtWidgets.QLabel(parent)\n\
+        self.message = FittedTextBrowser(parent)\n\
+        self.copy = QtWidgets.QPushButton(parent)\n\
+        self.buttons = QtWidgets.QDialogButtonBox(parent)\n\
+\n\
+        self.copy.setText('Copy To Clipboard')\n\
+\n\
+        self.layout.addWidget(self.icon, 0, 0)\n\
+        self.layout.addWidget(self.message, 0, 1, 1, 2)\n\
+        self.layout.addWidget(self.copy, 1, 1)\n\
+        self.layout.addWidget(self.buttons, 1, 2)\n\
+\n\
+        self.layout.setAlignment(self.copy, QtCore.Qt.AlignLeft)\n\
+\n\
+        self.layout.setRowStretch(0, 1)\n\
+        self.layout.setColumnStretch(1, 1)\n\
+\n\
+\n\
+class Dialog(QtWidgets.QDialog):\n\
+    def __init__(self, *args, cancellable=False, **kwargs):\n\
+        super().__init__(*args, **kwargs)\n\
+\n\
+        self.ui = DialogUi(parent=self)\n\
+\n\
+        self.ui.buttons.accepted.connect(self.accept)\n\
+        self.ui.buttons.rejected.connect(self.reject)\n\
+\n\
+        self.ui.copy.clicked.connect(self.copy)\n\
+\n\
+        self.setLayout(self.ui.layout)\n\
+        buttons = QtWidgets.QDialogButtonBox.Ok\n\
+        if cancellable:\n\
+            buttons |= QtWidgets.QDialogButtonBox.Cancel\n\
+\n\
+        self.ui.buttons.setStandardButtons(buttons)\n\
+\n\
+        self.text = None\n\
+        self.html = None\n\
+\n\
+        self.cached_maximum_size = self.maximumSize()\n\
+\n\
+        desktops = QtWidgets.QApplication.desktop()\n\
+        screen_number = desktops.screenNumber(self.parent())\n\
+        geometry = desktops.screenGeometry(screen_number)\n\
+\n\
+        self.setMaximumHeight(geometry.height() * 0.7)\n\
+        self.setMaximumWidth(geometry.width() * 0.7)\n\
+\n\
+    def copy(self):\n\
+        QtWidgets.QApplication.clipboard().setText(\n\
+            self.ui.message.toPlainText() + '\\n'\n\
+        )\n\
+\n\
+    def set_text(self, text):\n\
+        self.ui.message.setPlainText(text)\n\
+\n\
+        self.html = None\n\
+        self.text = text\n\
+\n\
+        self.ui.message.setLineWrapMode(QtWidgets.QTextEdit.NoWrap)\n\
+\n\
+    def set_html(self, html):\n\
+        self.ui.message.setHtml(html)\n\
+\n\
+        self.html = html\n\
+        self.text = None\n\
+\n\
+        self.ui.message.setLineWrapMode(QtWidgets.QTextEdit.WidgetWidth)\n\
+\n\
+    def set_message_box_icon(self, icon):\n\
+        self.ui.icon.setPixmap(QtWidgets.QMessageBox.standardIcon(icon))\n\
+\n\
+    def exec(self):\n\
+        QtCore.QTimer.singleShot(10, functools.partial(\n\
+                self.setMaximumSize,\n\
+                self.cached_maximum_size,\n\
+        ))\n\
+\n\
+        return super().exec()\n\
+\n\
+\n\
+def dialog(parent, message, title=None, icon=None,\n\
+           rich_text=False, cancellable=False):\n\
+    box = Dialog(parent=parent, cancellable=cancellable)\n\
+\n\
+    if rich_text:\n\
+        box.set_html(message)\n\
+    else:\n\
+        box.set_text(message)\n\
+\n\
+    if icon is not None:\n\
+        box.set_message_box_icon(icon)\n\
+\n\
+    if title is not None:\n\
+        parent_title = QtWidgets.QApplication.instance().applicationName()\n\
+\n\
+        if len(parent_title) > 0:\n\
+            title = ' - '.join((\n\
+                parent_title,\n\
+                title,\n\
+            ))\n\
+\n\
+\n\
+        box.setWindowTitle(title)\n\
+\n\
+    return box.exec()\n\
+\n\
+QtCore.qDebug('Setting sys.excepthook to provide dialog')\n\
+sys.excepthook = exception_message_box\n\
+";
+
+
 // Construct the collection of Python widgets.
 PyCustomWidgets::PyCustomWidgets(QObject *parent) : QObject(parent),
         sys_path(0), sip_unwrapinstance(0), qtdesigner_custom(0)
@@ -134,6 +310,107 @@
         PyGILState_STATE gil_state = PyGILState_Ensure();
 #endif
 
+        qDebug() << "Preparing to import module";
+
+        // https://stackoverflow.com/a/42853449/228539
+        PyObject *pyModule = PyModule_New(moduleName);
+        // Set properties on the new module object
+        PyModule_AddStringConstant(pyModule, "__file__", "");
+        PyObject *localDict = PyModule_GetDict(pyModule);   // Returns a borrowed reference: no need to Py_DECREF() it once we are done
+        PyObject *builtins = PyEval_GetBuiltins();  // Returns a borrowed reference: no need to Py_DECREF() it once we are done
+        PyDict_SetItemString(localDict, "__builtins__", builtins);
+
+        qDebug() << "About to import module";
+
+        // Define code in the newly created module
+        PyObject *pyValue = PyRun_String(moduleCode, Py_file_input, localDict, localDict);
+        if (pyValue == NULL)
+        {
+            qDebug() << "An error occurred while importing the module";
+
+            char * full_backtrace = NULL;
+            
+            if (PyErr_Occurred() != NULL) {
+                qDebug() << "PyErr_Occurred() != NULL";
+                PyObject *ptype, *pvalue, *ptraceback;
+                PyObject *pystr, *module_name, *pyth_module, *pyth_func;
+                char *str;
+                char *error_description;
+
+                PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+                qDebug() << "ptype == NULL : " << (ptype == NULL);
+                qDebug() << "pvalue == NULL : " << (pvalue == NULL);
+                qDebug() << "ptraceback == NULL : " << (ptraceback == NULL);
+                pystr = PyObject_Str(pvalue);
+                qDebug() << "pystr == NULL : " << (pystr == NULL);
+                pystr = PyUnicode_AsEncodedString(pystr, "utf-8", "error 1?");
+                qDebug() << "pystr == NULL : " << (pystr == NULL);
+                str = PyBytes_AS_STRING(pystr);
+                qDebug() << "str == NULL : " << (str == NULL);
+                if (str != NULL)
+                {
+                    qDebug() << "str: " << str;
+                }
+                Py_DECREF(pystr);
+                error_description = strdup(str);
+                qDebug() << "error_description == NULL : " << (error_description == NULL);
+                if (error_description != NULL)
+                {
+                    qDebug() << "error_description: " << error_description;
+                }
+
+                /* See if we can get a full traceback */
+                module_name = PyUnicode_Decode("traceback", 9, "utf-8", "error 2?");
+                pyth_module = PyImport_Import(module_name);
+                Py_DECREF(module_name);
+
+                qDebug() << "pyth_module == NULL : " << (pyth_module == NULL);
+                if (pyth_module == NULL) {
+                    full_backtrace = NULL;
+                    return;
+                }
+
+                pyth_func = PyObject_GetAttrString(pyth_module, "format_exception");
+                if (pyth_func && PyCallable_Check(pyth_func))
+                {
+                    qDebug() << "pyth_func && PyCallable_Check(pyth_func)";
+                    PyObject *pyth_val;
+
+                    pyth_val = PyObject_CallFunctionObjArgs(pyth_func, ptype, pvalue, ptraceback, Py_None, true);
+                    qDebug() << "pyth_val == NULL : " << (pyth_val == NULL);
+
+                    pystr = PyObject_Str(pyth_val);
+                    qDebug() << "pystr == NULL : " << (pystr == NULL);
+
+                    pystr = PyUnicode_AsEncodedString(pystr, "utf-8", "error 3?");
+                    qDebug() << "pystr == NULL : " << (pystr == NULL);
+
+                    str = PyBytes_AS_STRING(pystr);
+                    qDebug() << "str == NULL : " << (str == NULL);
+
+                    Py_DECREF(pystr);
+                    full_backtrace = strdup(str);
+                    Py_DECREF(pyth_val);
+                }
+            }
+            
+            if (full_backtrace != NULL)
+            {
+                qDebug() << "About to print backtrace";
+                qDebug() << full_backtrace;
+            }
+            else
+            {
+                qDebug() << "Backtrace was NULL";
+            }
+        }
+        else
+        {
+            Py_DECREF(pyValue);
+        }
+
+        qDebug() << "Done to importing module";
+
         bool fatal = importPlugins(dir, plugins);
 
 #if defined(WITH_THREAD)
