diff --git a/qmlscene/pluginloader.cpp b/qmlscene/pluginloader.cpp
index 635d31e..45e3682 100644
--- a/qmlscene/pluginloader.cpp
+++ b/qmlscene/pluginloader.cpp
@@ -21,6 +21,7 @@
  */
 
 
+#include <iostream>
 #include <stdlib.h>
 
 #include <Python.h>
@@ -218,6 +219,7 @@ void PyQt5QmlPlugin::registerTypes(const char *uri)
         return;
 
 #ifdef WITH_THREAD
+    std::cerr << "pluginloader.cpp debug: PyGILState_Ensure() (A)\n";
     PyGILState_STATE gil = PyGILState_Ensure();
 #endif
 
@@ -225,6 +227,7 @@ void PyQt5QmlPlugin::registerTypes(const char *uri)
         PyErr_Print();
 
 #ifdef WITH_THREAD
+    std::cerr << "pluginloader.cpp debug: PyGILState_Release() (A)\n";
     PyGILState_Release(gil);
 #endif
 }
@@ -268,9 +271,22 @@ bool PyQt5QmlPlugin::callRegisterTypes(const QString &py_plugin,
 
     if (!plugin_type)
     {
+        const char * py_plugin_char_pointer = py_plugin.toLatin1().constData();
+        std::cerr << "pluginloader.cpp debug: (void *) py_plugin_char_pointer " << (void *) py_plugin_char_pointer << "\n";
+        std::cerr << "pluginloader.cpp debug: py_plugin_char_pointer bytes";
+        int length = 0;
+        for (int i = 0; i < 100; i++) {
+            int b = py_plugin_char_pointer[i];
+            std::cerr << " " << b;
+            if (b == 0) {
+                length = i;
+                break;
+            }
+        }
+        std::cerr << "pluginloader.cpp debug: py_plugin_char_pointer 0 byte at index " << length;
         PyErr_Format(PyExc_AttributeError,
                 "%s does not contain an implementation of QQmlExtensionPlugin",
-                py_plugin.toLatin1().constData());
+                py_plugin_char_pointer);
 
         Py_DECREF(plugin_mod);
         return false;
@@ -295,6 +311,7 @@ bool PyQt5QmlPlugin::callRegisterTypes(const QString &py_plugin,
         if (res_obj)
         {
 #if PY_MAJOR_VERSION >= 3
+            std::cerr << "pluginloader.cpp debug: res_obj " << res_obj << "\n";
             PyErr_Format(PyExc_TypeError,
                     "unexpected result from registerTypes(): %S", res_obj);
 #else
@@ -412,6 +429,16 @@ PyObject *PyQt5QmlPlugin::getModuleAttr(const char *module, const char *attr)
 void PyQt5QmlPlugin::getSipAPI()
 {
 #if defined(SIP_USE_PYCAPSULE)
+    PyObject *mod = PyImport_ImportModule("PyQt5.sip");
+    if (!mod)
+    {
+        PyErr_Print();
+        return;
+    }
+    else
+    {
+        Py_DECREF(mod);
+    }
     sip = (const sipAPIDef *)PyCapsule_Import("PyQt5.sip._C_API", 0);
 #else
     PyObject *c_api = getModuleAttr("PyQt5.sip", "_C_API");
